Final updates for next release

paulromano opened the pull request:
This pull request adds release notes for version 0.10.0, fills in missing documentation for the C API, adds a C/C++ header file (include/openmc.h), and makes a few minor fixes (mostly doc fixes). Once this is merged, I'll tag a new release.

paulromano commented:
Any takers to review? @smharper @nelsonag

nelsonag commented:
I'll probably be able to get to it this weekend

On Jan 5, 2018 1:15 PM, "Paul Romano" <notifications@github.com> wrote:

> Any takers to review? @smharper <https://github.com/smharper> @nelsonag
> <https://github.com/nelsonag>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/950#issuecomment-355625840>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_TM1VX1TKOoFqLbsevlt3QbjBkuXbbks5tHmbBgaJpZM4RQMZ_>
> .
>

nelsonag commented on docs/source/usersguide/install.rst:
```diff
@@ -113,34 +113,32 @@ Prerequisites
+      OpenMC uses HDF5 for many input/output files. As such, you will need to
+      have HDF5 installed on your computer. The installed version will need to
+      have been compiled with the same compiler you intend to compile OpenMC
+      with. On Debian derivatives, HDF5 and/or parallel HDF5 can be installed
```
'If compiling with gcc from the repositories, users of Debian derivatives can...'

nelsonag commented on docs/source/usersguide/install.rst:
```diff
@@ -113,34 +113,32 @@ Prerequisites
+      .. code-block:: sh
 
-      You may omit ``--enable-parallel`` if you want to compile HDF5_ in serial.
+          sudo apt install libhdf5-dev
```
Why cant the user install libhdf5-openmpi-dev (or mpich2) if they want hdf5 parallel support?  The docs seem to read like "if you want parallel support, you have to compile it yourself."   Also, do we not need hdf5-helpers?

paulromano replied:
I realized that libhdf5-dev depends on hdf5-helpers, so it's not necessary to list that one too. Good point regarding the parallel versions of libhdf5-dev. I'll update this.

paulromano commented on docs/source/capi/index.rst:
```diff
@@ -8,28 +8,132 @@ C API
+
+   :param int32_t index: Index in the filters array
+   :param double** energies: Bounding energies of the bins for the energy filter
+   :param int32_t* n: Number of energies specified
```
Unfortunately in a language without exceptions (C and Fortran), one is left resorting to methods like what I've used here. Note that throughout the C API, return values consistently are used to indicate errors, and our Python bindings take advantage of this when checking for errors (non-zero return statuses get turned into exceptions in Python). When a value needs to be returned by a function in the C API, pointer arguments have to be used.

nelsonag commented:
I'm with you on that one.  My only point is that this scheme inherently
makes it not obvious which parameters are for returns and which are input.
Just asking for clarification in the docs on that.

On Jan 6, 2018 3:36 PM, "Paul Romano" <notifications@github.com> wrote:

> *@paulromano* commented on this pull request.
> ------------------------------
>
> In docs/source/capi/index.rst
> <https://github.com/mit-crpg/openmc/pull/950#discussion_r160034341>:
>
> > -                    instances, pass a null pointer.
> -   :type instance: int32_t*
> +   :param int32_t index: Index in the cells array
> +   :param double T: Temperature in Kelvin
> +   :param int32_t* instance: Which instance of the cell. To set the temperature
> +                             for all instances, pass a null pointer.
> +   :return: Return status (negative if an error occurred)
> +   :rtype: int
> +
> +.. c:function:: int openmc_energy_filter_get_bins(int32_t index, double** energies, int32_t* n)
> +
> +   Return the bounding energies for an energy filter
> +
> +   :param int32_t index: Index in the filters array
> +   :param double** energies: Bounding energies of the bins for the energy filter
> +   :param int32_t* n: Number of energies specified
>
> Unfortunately in a language without exceptions (C and Fortran), one is
> left resorting to methods like what I've used here. Note that throughout
> the C API, return values consistently are used to indicate errors, and our
> Python bindings take advantage of this when checking for errors (non-zero
> return statuses get turned into exceptions in Python). When a value needs
> to be returned by a function in the C API, pointer arguments have to be
> used.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/950#discussion_r160034341>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_TMyCnbHegDrL0ZlXc5gJX6hMJ6fm_ks5tH9lJgaJpZM4RQMZ_>
> .
>

paulromano commented on docs/source/devguide/docbuild.rst:
```diff
@@ -35,21 +36,9 @@ Building Documentation as a PDF
      make latexpdf
-
-.. _Sphinx: http://sphinx-doc.org
-.. _sphinxcontrib-tikz: https://bitbucket.org/philexander/tikz
```
Nope, there are two TeX files that need to be converted, but we also just store the PNGs directly so that someone building the docs doesn't need to have that package,

paulromano commented on docs/source/usersguide/install.rst:
```diff
@@ -95,10 +95,10 @@ Prerequisites
-      compiler using the following command::
+      OpenMC includes various source files written in C and C++,
+      respectively. These source files have been tested to work with a wide
+      variety of compilers. If you are using a Debian-based distribution, you
```
Not really that worried about it. I haven't heard of anyone trying the PGI compiler other than myself.

paulromano commented:
@nelsonag I believe I've addressed all the issues you raised. With respect to arguments being input/output, I've marked any pointers in C functions which are not intended to be output as `const` so that it is clearly indicated to a user.

As for the comment about arrays of pointers versus pointer to pointer, I've changed most of them to ** for consistency.
